File: decrease_conquer/hakjum_cal.py
import os

BASE_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), 'data')

class Graduate:
    def __init__(self):
        global BASE_DIR
        self.gyojic=[]
        self.major=[]
        with open(os.path.join(BASE_DIR, 'gyoil.txt'), 'r') as f:
            gyoil = list(map(lambda n:n[:-1], f.readlines()))
            self.gyojic.append(gyoil)
        with open(os.path.join(BASE_DIR, 'gyopil.txt'), 'r') as f:
            gyopil = list(map(lambda n:n[:-1], f.readlines()))
            self.gyojic.append(gyopil)
        self.gyo_unit = 22
        self.gyo_nowunit = (len(self.gyojic[0])+len(self.gyojic[1])-2)*2

        while True:
            print("단일/복수전공 여부를 입력해주세요")
            print("1은 단일전공, 2는 복수전공입니다.")
            select = input(">> ")
            if select=="1" or select=="2":
                self.major_unit = (66 if select=="1" else 50)
                break
            else:
                print("잘못 입력하셨습니다. 다시 입력해주세요.")

        while True:
            print("학번을 입력해주세요.")
            print("예를 들어, 18학번은 18을 입력해주세요.")
            try:
                self.hakbun = int(input(">> "))
                self.txt = (['jeonil_old.txt', 'jeonpil_old.txt'] if self.hakbun<17 else ['jeonil_new.txt', 'jeonpil_new.txt'])
                for t in self.txt:
                    with open(os.path.join(BASE_DIR, t), 'r') as f:
                        tmp = list(map(lambda n:n[:-1], f.readlines()))
                        self.major.append(tmp)
                check = self.major[0]+self.major[1]
                self.major_nowunit = self.major_unit - (32-len(check))*3 + (0 if "상업정보교과논리논술" in check else 1)
                break
            except ValueError:
                print("잘못 입력하셨습니다. 다시 입력해주세요.")
                
    def status(self):
        print("="*15)
        print("남은 학점")
        print("   전공 {}/{}".format(self.major_nowunit, self.major_unit))
        print("   교직 {}/{}\n".format(self.gyo_nowunit, self.gyo_unit))
        print("남은 전공필수 과목")
        for lesson in self.major[1]:
            print("   {}".format(lesson))
        print("남은 교직필수 과목")
        for lesson in self.gyojic[1]:
            print("   {}".format(lesson))
        print("="*15)
        
    def listen_lesson(self, lesson):
        for g in self.gyojic:
            if lesson in g:
                g.remove(lesson)
                self.gyo_nowunit -= 2
                if self.gyo_nowunit<0:
                    self.gyo_nowunit = 0
                self.save_gyo()
                return True
        for m in self.major:
            if lesson in m:
                m.remove(lesson)
                self.major_nowunit -= 3
                if lesson == "상업정보교과논리논술":
                    self.major_nowunit += 1
                if self.major_nowunit<0:
                    self.major_nowunit=0
                self.save_major()
                return True
        return False
    
    def save_gyo(self):
        os.remove(os.path.join(BASE_DIR, 'gyoil.txt'))
        os.remove(os.path.join(BASE_DIR, 'gyopil.txt'))
        with open(os.path.join(BASE_DIR, 'gyoil.txt'), 'w') as f:
            f.writelines(list(map(lambda n:n+'\n', self.gyojic[0])))
        with open(os.path.join(BASE_DIR, 'gyopil.txt'), 'w') as f:
            f.writelines(list(map(lambda n:n+'\n', self.gyojic[1])))
            
    def save_major(self):
        for i in range(2):
            os.remove(os.path.join(BASE_DIR, self.txt[i]))
            with open(os.path.join(BASE_DIR, self.txt[i]), 'w') as f:
                f.writelines(list(map(lambda n:n+'\n', self.major[i])))
            
    def is_graduate(self):
        return self.gyo_nowunit<=0 and self.major_nowunit<=0 and not self.major[1] and not self.gyojic[1]
                    
# 본 프로그램 부분

now = Graduate()
now.status()
while True:
    cmd = input(" >> ")
    if cmd=='status' or now.listen_lesson(cmd):
        now.status()
        if now.is_graduate():
            print("졸업!!!")
            break
    else:
        print("잘못 입력하셨습니다. 다시 입력해주세요.")
        print("입력 가능한 문자열은 문제 조건을 확인해주세요.")
        

File: decrease_conquer/hakjum_cal.cpp
#include <iostream>
#include <locale>
#include <fstream>
#include <vector>
#include <string>

using namespace std;

// TODO : UTF-8 인코딩 형식의 txt파일을 읽어오고 쓰는 과정을 구현하지 못했습니다 ㅠㅠ...
// 이 코드에는 최대한 시도한 흔적을 남겨보았습니다.
// 파이썬 파일의 경우에는 파일 입출력이 구현되어있습니다.
// 이후 재재출 기회가 생긴다면 꼭 구현해보겠습니다!!

class Graduate{
private:
    vector<string> mlist;
    vector<string> glist {"../data/gyoil.txt", "../data/gyopil.txt"};
    vector<vector<string>> gyojic, major;
    int gyo_unit=22, gyo_nowunit, major_unit, major_nowunit, hakbun;
    void readFile(string file, vector<string> &at){
        wstring buff;
        wfstream fs(file);
        while ( getline(fs, buff) ) {
            string tmp;
            tmp.assign(buff.begin(), buff.end());
            if(tmp=="") break;
            at.push_back(tmp);
        }
        fs.close();
    }

public:
    Graduate(){
        for(string g : glist){
            vector<string> tmp;
            readFile(g, tmp);
            gyojic.push_back(tmp);
        }
        gyo_nowunit = ((int)gyojic[0].size()+(int)gyojic[1].size()-2)*2;

        while(true){
            cout << "단일/복수전공 여부를 입력해주세요" << endl << "1은 단일전공, 2는 복수전공입니다." << endl << ">> ";
            int select;
            cin >> select;
            if(select==1){
                major_unit = 66; break;
            } else if (select==2){
                major_unit = 50; break;
            } else{
                cout << "잘못 입력하셨습니다. 다시 입력해주세요." << endl;
            }
        }

        cout << "학번을 입력해주세요." << endl << "예를 들어, 18학번은 18을 입력해주세요." << endl << ">> ";
        cin >> hakbun;
        if(hakbun<17){
            mlist.push_back("../data/jeonil_old.txt");
            mlist.push_back("../data/jeonpil_old.txt");
        } else{
            mlist.push_back("../data/jeonil_new.txt");
            mlist.push_back("../data/jeonpil_new.txt");
        }

        for(auto m : mlist){
            vector<string> tmp;
            ifstream in(m);
            string s;
            while (in) {
                getline(in, s);
                tmp.push_back(s);
            }
            in.close();
            major.push_back(tmp);
        }
        major_nowunit = major_unit - (32-(int)mlist[0].size()-(int)mlist[1].size())*3+1;
        for(auto mp : major[1]){
            if(mp=="상업정보교과논리논술"){
                major_nowunit--;
                break;
            }
        }
    }

    void status(){
        cout << "===============\n남은 학점" << endl;
        cout << "   전공 " << major_nowunit << "/" << major_unit << endl;
        cout << "   교직 " << gyo_nowunit << "/" << gyo_unit << "\n\n";
        cout << "남은 전공필수 과목" << endl;
        for(auto m : major){
            for(auto lesson : m) cout << "   " << lesson << endl;
        }

        cout << "남은 교직필수 과목" << endl;
        for(auto lesson : gyojic[1]) cout << "   " << lesson << endl;
        cout << "===============" << endl;
    }

    bool listen_lesson(string lesson) {
        return false;
    }

    bool is_graduate(){
        return gyo_nowunit<=0 && major_nowunit<=0 && major[1].empty() && gyojic[1].empty();
    }
};

int main(){
    locale::global(locale("ko_KR.UTF-8")); // 맨 처음 한번 실행
    Graduate now;
    now.status();
    while (true){
        string cmd;
        cout << ">> ";
        cin.clear();
        getline(cin, cmd);
        if(cmd=="status" || now.listen_lesson(cmd)){
            now.status();
            if(now.is_graduate()){
                cout << "졸업!!!";
                return 0;
            }
        } else cout << "잘못 입력하셨습니다. 다시 입력해주세요." << endl << "입력 가능한 문자열은 문제 조건을 확인해주세요." << endl;
    }
}